fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/ContainmentIndexImpl.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to DuraSpace under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * DuraSpace licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fcrepo.kernel.impl;
+
+import org.fcrepo.kernel.api.ContainmentIndex;
+import org.fcrepo.kernel.api.Transaction;
+import org.fcrepo.kernel.api.models.FedoraResource;
+import org.springframework.jdbc.datasource.DriverManagerDataSource;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import java.sql.SQLException;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.util.stream.Stream;
+
+/**
+ * @author peichman
+ */
+public class ContainmentIndexImpl implements ContainmentIndex {
+    @Inject
+    private DriverManagerDataSource containmentIndexDataSource;
+
+    private Connection conn;
+
+    private PreparedStatement childrenQuery;
+
+    private static final String FEDORA_ID_COLUMN = "fedoraId";
+
+    private static final String SELECT_CHILDREN = "SELECT " + FEDORA_ID_COLUMN + " FROM resources WHERE parent = ?";
+
+    /**
+     * Connect to the database
+     */
+    @PostConstruct
+    private void setup() {
+        try {
+            conn = containmentIndexDataSource.getConnection();
+            childrenQuery = conn.prepareStatement(SELECT_CHILDREN);
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Do the actual database query to get the contained IDs.
+     *
+     * @param fedoraId ID of the containing resource
+     * @param transactionId ID of the current transaction (if any)
+     * @return A stream of contained identifiers
+     */
+    private Stream<String> getChildren(final String fedoraId, final String transactionId) {
```

Comment:
Do we also need a method to update/insert into the containment index?

Reply:
Yes. I was looking into where to hook that in, and was thinking something in the internal event bus (same place that generates the JMS messages).

Reply:
Started in https://github.com/fcrepo4/fcrepo4/pull/1584/commits/e294e8cb14d605f6ae347f38d9601336982a0a01

Reply:
Added the remove counterpart in https://github.com/fcrepo4/fcrepo4/pull/1584/commits/dcd44266eee05e576ba7417d066bdab0aeb46f00